libsolidity/analysis/StaticAnalyzer.cpp

```diff
@@ -346,24 +346,19 @@ bigint StaticAnalyzer::structureSizeEstimate(Type const& _type, set<StructDefini
 	case Type::Category::Array:
 	{
 		auto const& t = dynamic_cast<ArrayType const&>(_type);
-		return structureSizeEstimate(*t.baseType(), _structsSeen) * (t.isDynamicallySized() ? 1 : t.length());
+		return t.isDynamicallySized() ? bigint(1) : (structureSizeEstimate(*t.baseType(), _structsSeen) * t.length());
 	}
 	case Type::Category::Struct:
 	{
 		auto const& t = dynamic_cast<StructType const&>(_type);
+		solAssert(!_structsSeen.count(&t.structDefinition()), "Recursive struct.");
 		bigint size = 1;
-		if (!_structsSeen.count(&t.structDefinition()))
-		{
-			_structsSeen.insert(&t.structDefinition());
-			for (auto const& m: t.members(nullptr))
-				size += structureSizeEstimate(*m.type, _structsSeen);
-		}
+		_structsSeen.insert(&t.structDefinition());
+		for (auto const& m: t.members(nullptr))
+			size += structureSizeEstimate(*m.type, _structsSeen);
+		_structsSeen.erase(&t.structDefinition());
 		return size;
 	}
-	case Type::Category::Mapping:
-	{
-		return structureSizeEstimate(*dynamic_cast<MappingType const&>(_type).valueType(), _structsSeen);
-	}
```

Comment:
There will still be one slot for them, won't there? Maybe better to have this be
```
	case Type::Category::Mapping:
	{
		return 1;
	}
```

Reply:
Sure, but `1` is the default value:

```
    default:
		break;
	}
	return bigint(1);
}
```

First I did it the way you suggested :), but then decided to remove the `case Type::Category::Mapping`, keeping in mind upcoming `StaticAnalyzer::structureSizeEstimate` -> `Types::storageSizeEstimate()` refactring.